crates/macros/impl/src/cheatcodes.rs

```diff
@@ -0,0 +1,252 @@
+                _ => return Err(meta.error("unexpected attribute")),
+            };
+            Ok(())
+        })?;
+    }
+    let group = group.ok_or_else(|| {
+        syn::Error::new(name.span(), "missing #[cheatcode(group = ...)] attribute")
+    })?;
+    let status = status.unwrap_or_else(|| Ident::new("Stable", Span::call_site()));
+    let safety = if let Some(safety) = safety {
+        quote!(Safety::#safety)
+    } else {
+        let panic = quote_spanned! {name.span()=>
+            panic!("cannot determine safety from the group, add a `#[cheatcode(safety = ...)]` attribute")
+        };
+        quote! {
+            match Group::#group.safety() {
+                Some(s) => s,
+                None => #panic,
+            }
+        }
+    };
+
+    check_named_fields(data, name);
+
+    let id = name.to_string();
+    let id = id.strip_suffix("Call").expect("function struct ends in Call");
+
+    let doc = get_docstring(attrs);
+    let (signature, selector, declaration, description) = func_docstring(&doc);
+
+    let (visibility, mutability) = parse_function_attrs(declaration, name.span())?;
+    let visibility = Ident::new(visibility, Span::call_site());
+    let mutability = Ident::new(mutability, Span::call_site());
+
+    if description.is_empty() {
+        emit_warning!(name.span(), "missing documentation for a cheatcode")
+    }
+    let description = description.replace("\n ", "\n");
+
+    Ok(quote! {
+        impl CheatcodeDef for #name {
+            const CHEATCODE: &'static Cheatcode<'static> = &Cheatcode {
+                id: #id,
+                declaration: #declaration,
+                visibility: Visibility::#visibility,
+                mutability: Mutability::#mutability,
+                signature: #signature,
+                selector: #selector,
+                description: #description,
+                group: Group::#group,
+                status: Status::#status,
+                safety: #safety,
+            };
+        }
+    })
+}
+
+/// Generates the `CHEATCODES` constant and implements `CheatcodeImpl` dispatch for an enum.
+fn derive_enum(name: &Ident, input: &syn::DataEnum) -> Result<TokenStream> {
+    if input.variants.iter().any(|v| v.fields.len() != 1) {
+        return Err(syn::Error::new(name.span(), "expected all variants to have a single field"))
+    }
+
+    // keep original order for matching
+    let variants_names = input.variants.iter().map(|v| &v.ident);
+
+    let mut variants = input.variants.iter().collect::<Vec<_>>();
+    variants.sort_by(|a, b| a.ident.cmp(&b.ident));
+    let variant_tys = variants.iter().map(|v| {
+        assert_eq!(v.fields.len(), 1);
+        &v.fields.iter().next().unwrap().ty
+    });
+    Ok(quote! {
+        /// All the cheatcodes in [this contract](self).
+        pub const CHEATCODES: &'static [&'static Cheatcode<'static>] = &[#(<#variant_tys as CheatcodeDef>::CHEATCODE,)*];
```

Comment:
is there some type magic behind why this is `const &'static`?

Reply:
Not sure I understand the question. You have to specify `'static` lifetimes and you can only have `const`s in impl blocks